extensions/amp-ad-network-doubleclick-impl/0.1/amp-ad-network-doubleclick-impl.js

```diff
@@ -715,6 +715,26 @@ export class AmpAdNetworkDoubleclickImpl extends AmpA4A {
     return {'artc': artc.join() || null, 'ati': ati.join(), 'ard': ard.join()};
   }
 
+  /** @override */
+  getCustomRealTimeConfigMacros_() {
+    return {
+      PAGEVIEWID: () => {
+        const docInfo = Services.documentInfoForDoc(this.element);
+        return docInfo.pageViewId;
+      },
+      HREF: () => this.win.location.href,
+      DATASLOT: () => this.element.getAttribute('data-slot'),
```

Comment:
Consider whitelist of allowed attributes and then single macro with parameter of name of attribute to be extracted:

const whitelist = {};
ATTR = name => !!whitelist[name] && this.element.getAttribute(name)

Reply:
This doesn't work with how the url replacements works. It needs a full list of all valid macros to look for in the URL. This suggestion would only work if we know what the valid macros are already, and then go from there.